### PR TITLE
feat(mimo): Phase 4 - MIMO training, checkpointing, and evaluation infrastructure

### PR DESCRIPTION
**Pull request by aroshanghias-nvd**
## Summary

Consolidated MIMO training infrastructure (Phases 4 & 5) for heterogeneous multi-module models:

### Phase 4: Training Loop
- **mimo_parallel_utils.py**: Multimodule gradient sync, `MultiModuleProcessGroupCollection` builder, stub rank validation, data loader contract checks
- **mimo_step.py**: Forward step and loss functions with dict-based tensor flow for pipeline parallelism
- **train_mimo.py**: Main training loop using `MultiModulePipelineCommunicator` and `MimoOptimizer`
- **pretrain_mimo.py**: Entry point with `setup_mimo()` composite helper

### Phase 5: Checkpointing & Evaluation
- **train_mimo.py**: Integrated `save_checkpoint` with `MimoOptimizer`, evaluation via `evaluate_and_print_results` with multimodule communicator, async checkpoint finalization with `maybe_finalize_async_save`
- **dp_utils.py**: Added `slice_batch_for_mimo` for DP-aware batch slicing
- **mimo_config.py**: Added parallelism constraint validation (TP power-of-2, DP pairwise divisibility, encoder DP >= LLM DP)
- **mimo_provider.py**: Added `unwrap_mimo_model()`, provider defaults resolution, pipeline stage handling in `forward_step()`
- **eval.py**: Extended `evaluate()` and `evaluate_and_print_results()` with injectable `p2p_communicator` and `pg_collection` for multimodule support

### Architectural Decisions (from PR review)
- `initialize_model_parallel()` raises `NotImplementedError` — MIMO uses per-module HyperCommGrids, not global MPU state
- Removed misleading `tensor/pipeline/context_parallel_size` proxy properties — MIMO parallelism is per-module, not global
- Topology is a configurable field on `MimoModelProvider` with auto-derivation fallback
- `MimoParallelismConfig.finalize()` requires `world_size: int` (not Optional) — MIMO always runs in a distributed environment

## Files Changed

| File | Description |
|------|-------------|
| mimo_parallel_utils.py | Multimodule gradient sync, PG collection builder |
| mimo_step.py | Forward step and loss functions |
| train_mimo.py | Main training loop with checkpointing, evaluation, profiling |
| pretrain_mimo.py | Entry point with setup_mimo() |
| dp_utils.py | slice_batch_for_mimo for DP-aware batch slicing |
| mimo_config.py | Parallelism constraint validation |
| mimo_provider.py | Provider defaults, unwrap utility, topology field |
| mimo_builder.py | Removed _default_topology (now configurable) |
| eval.py | Multimodule evaluation support |
| test_mimo_training_e2e.py | E2E training test with configurable parallelism |
| run_mimo_parallelism_tests.sh | Runner with 8 predefined parallelism configs |

## Dependencies

- Requires mimo/phase3-data-loading to be merged first
- Requires megatron-core with PR 3129 (MultiModulePipelineCommunicator, MultiModuleProcessGroupCollection)

## Test Plan

- [x] Unit tests for `mimo_parallel_utils.py` (validation, context managers)
- [x] Unit tests for `mimo_step.py` (loss_func, get_batch, forward_step)
- [x] Unit tests for `mimo_config.py` (parallelism constraints, finalize)
- [x] Unit tests for `test_mimo_provider.py` (provider, build_infra, DDP)
- [x] E2E training test with configurable TP/PP/DP parallelism
- [ ] Integration test with actual multi-GPU heterogeneous setup

> **Note**: This PR consolidates the former PR #2182 (Phase 5) into this PR for streamlined review.

### CODE REVIEW

File: src/megatron/bridge/models/mimo/mimo_provider.py [hunk 1]
```diff
@@ -0,0 +1,540 @@
...
+                    for hook in pre_wrap_hook:
+                        result = hook(model)
+                        if result is not None:
+                            model = result
+                    return model
+                return composed_hook
+            return pre_wrap_hook
+        return self.pre_wrap_hook
+    
+    def initialize_model_parallel(
```
Comment (yashaswikarnati) [inline src/megatron/bridge/models/mimo/mimo_provider.py:488]: do we need this method?

File: src/megatron/bridge/models/mimo/mimo_provider.py [hunk 2]
```diff
@@ -0,0 +1,540 @@
...
+    # Fields required by ModelProviderMixin / get_model()
+    # These have sensible defaults for MIMO
+    fp16: bool = False
+    bf16: bool = True
+    use_cpu_initialization: bool = False
+    init_model_with_meta_device: bool = False
+    virtual_pipeline_model_parallel_size: Optional[int] = None
+    
+    @property
+    def tensor_model_parallel_size(self) -> int:
```
Comment (yashaswikarnati) [inline src/megatron/bridge/models/mimo/mimo_provider.py]: do we need all the tp/pp size methods here?

File: src/megatron/bridge/models/mimo/mimo_builder.py [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+            _ = grid.create_pg([dim])
+        # Create dp_cp composite group for gradient reduction
+        _ = grid.create_pg(["dp", "cp"])
+
+        grids[module_name] = grid
+
+    return grids
+
+
+def _default_topology(mimo_parallelism_config: MimoParallelismConfig) -> Dict[str, List[str]]:
```
Comment (yashaswikarnati) [inline src/megatron/bridge/models/mimo/mimo_builder.py]: just curious why do we need this

File: src/megatron/bridge/models/mimo/mimo_config.py [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+        )
+
+    @property
+    def total_ranks(self) -> int:
+        if self.data_parallel_size is None:
+            raise ValueError("data_parallel_size must be set before accessing total_ranks.")
+        return self.total_model_parallel_size * self.data_parallel_size
+
+    def finalize(self, world_size: Optional[int]) -> None:
+        """Compute data_parallel_size if unset, and validate parallelism constraints."""
```
Comment (yashaswikarnati) [inline src/megatron/bridge/models/mimo/mimo_config.py:36]: qq; when can the world size be None

File: src/megatron/bridge/models/mimo/mimo_provider.py (hunk 1 above)
Comment (aroshanghias-nvd) [reply src/megatron/bridge/models/mimo/mimo_provider.py:488]: No current MIMO code path calls `initialize_model_parallel()`. The override is purely defensive — it prevents someone from accidentally calling `mimo_provider.initialize_model_parallel()` in the future and silently corrupting process groups (because it will set global parallelism settings). 

But maybe it's best to be even more defensive, and raise NotImplementedError. Let me do that.

File: src/megatron/bridge/models/mimo/mimo_provider.py (hunk 2 above)
Comment (aroshanghias-nvd) [reply src/megatron/bridge/models/mimo/mimo_provider.py]: Good callout. Not needed, and may become an issue later, if not used properly. I removed them.

File: src/megatron/bridge/models/mimo/mimo_builder.py (hunk 3 above)
Comment (aroshanghias-nvd) [reply src/megatron/bridge/models/mimo/mimo_builder.py]: I've now replaced it with an explicit topology field on MimoModelProvider. When None (the default), the same encoder→llm topology is auto-derived from modality_submodules_spec. Setting it explicitly enables non-standard topologies (e.g., llm → generator) without changing any builder code. The _default_topology helper in mimo_builder.py is removed.

File: src/megatron/bridge/models/mimo/mimo_config.py (hunk 4 above)
Comment (aroshanghias-nvd) [reply src/megatron/bridge/models/mimo/mimo_config.py:36]: Good catch — it couldn't be None in any real MIMO run (distributed is always initialized), so this was just overly defensive for testability. I've tightened it: MimoParallelismConfig.finalize() now requires world_size: int, and the caller in MimoModelProvider.finalize() raises a RuntimeError if dist isn't initialized. The world-size cross-check now always runs.
